keras_nlp/samplers/contrastive_sampler_test.py

```diff
@@ -0,0 +1,141 @@
+        self.int_lookup = {i: chr(i + ord("a")) for i in range(26)}
+        self.char_lookup = {v: k for k, v in self.int_lookup.items()}
+        self.batch_size = 1
+        self.length = 12
+        self.hidden_dim = 3
+        self.vocab_size = len(self.int_lookup)
+        self.hidden_states = tf.ones(
+            [
+                self.batch_size,
+                self.length,
+                self.hidden_dim,
+            ]
+        )
+
+        def next(prompt, cache, index):
+            batch_size = tf.shape(prompt)[0]
+            # Return a distribution favoring the next char in cache.
+            logits = tf.one_hot(cache[:, index], self.vocab_size) * 1e9
+            hidden_states = tf.ones([batch_size, 1, self.hidden_dim])
+            return logits, hidden_states, cache
+
+        self.next = next
+        self.sampler = ContrastiveSampler(k=5, alpha=0.2)
+
+    def join_as_string(self, x):
+        return ["".join([self.int_lookup[i] for i in s]) for s in x.numpy()]
+
+    def test_stateless_call(self):
+        def next(prompt, cache, index):
+            # Return a distribution favoring the first token in the vocab.
+            batch_size = tf.shape(prompt)[0]
+            logits = (
+                tf.one_hot(
+                    tf.zeros(batch_size, dtype=tf.int32),
+                    self.vocab_size,
+                )
+                * 1e9
+            )
+            hidden_states = tf.ones([batch_size, 1, self.hidden_dim])
+            return logits, hidden_states, cache
+
+        prompt = tf.fill((self.batch_size, self.length), self.char_lookup["z"])
+        output = self.sampler(
+            next=next,
+            prompt=prompt,
+            index=5,
+            hidden_states=self.hidden_states,
+        )
+        self.assertEqual(self.join_as_string(output), ["zzzzzaaaaaaa"])
+
+    def test_stateful_call(self):
+        cache_chars = list("sequentiallyy")
+        cache = tf.constant([[self.char_lookup[c] for c in cache_chars]])
+        prompt = tf.fill((self.batch_size, self.length), self.char_lookup["s"])
+        output = self.sampler(
+            next=self.next,
+            prompt=prompt,
+            cache=cache,
+            index=1,
+            hidden_states=self.hidden_states,
+        )
+        self.assertEqual(self.join_as_string(output), ["sequentially"])
+
+    def test_early_stopping(self):
+        cache_chars = list("sequentiallyy")
+        cache = tf.constant([[self.char_lookup[c] for c in cache_chars]])
+        prompt = tf.fill((self.batch_size, self.length), self.char_lookup["s"])
+        output = self.sampler(
+            next=self.next,
+            prompt=prompt,
+            cache=cache,
+            end_token_id=self.char_lookup["t"],
+            index=0,
+            hidden_states=self.hidden_states,
+        )
+        self.assertEqual(self.join_as_string(output), ["sequentsssss"])
+
+    def test_outputs_in_top_k(self):
```

Comment:
is there any clever (and not overly long) testing we can do to assert the degeneration penalty is working right?

something like a huge penalty alpha, uniform distribution over our lowercase alphabet, and assert that the character order is `"abcdefgh"` instead of `"aaaaaaaa"`? would need to tinker with a hidden representation so `similarity(hidden('a'), hidden('a')) > similarity(hidden('a'), hidden('b'))`

Reply:
Found a way to penalize the best token, please check it out.